File: 0063-Unique-Paths-II/lc0063.cc
// C++ solution.

class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        if (obstacleGrid.empty()) { return 0; }
        int m = obstacleGrid.size();
        int n = obstacleGrid.front().size();

        // 1-d dp array is enough
        vector<int> dp(n, 0);
        dp[0] = 1;
        for (int i = 0; i < m; i++) {
            dp[0] &= (1 - obstacleGrid[i][0]);
            for (int j = 1; j < n; j++) {
                dp[j] = obstacleGrid[i][j] == 1 ? 0 : (dp[j-1] + dp[j]);
            }
        }
        return dp[n-1];
    }
};

File: 0063-Unique-Paths-II/lc0063.rs
// Rust solution.

impl Solution {
    pub fn unique_paths_with_obstacles(obstacle_grid: Vec<Vec<i32>>) -> i32 {
        if obstacle_grid.is_empty() { 0 }
        else {
            let (m, n) = (obstacle_grid.len(), obstacle_grid[0].len());
            let mut dp: Vec<i32> = vec![0; n];
            dp[0] = 1;
            for i in 0..m {
                dp[0] *= (1-obstacle_grid[i][0]);
                for j in 1..n {
                    dp[j] = (dp[j] + dp[j-1]) * (1-obstacle_grid[i][j]);
                }
            }
            dp[n-1]
        }
    }
}
